src/core/handle.ml

```diff
@@ -185,7 +199,33 @@ let handle_cmd : sig_state -> p_command -> sig_state * proof_data option =
       (* Actually add the symbol to the signature. *)
       out 3 "(symb) %s ≔ %a\n" x.elt pp_term t;
       let d = if op then None else Some(t) in
-      let ss = Sig_state.add_symbol ss e Defin x a impl d in
+      (fst (Sig_state.add_symbol ss e Defin x a impl d), None)
+  | P_inductive(e, id, a, l)     ->
+      (* Add the inductive type in the signature *)
+      let (ss, sym_typ) = handle_symbol ss e Defin id [] a in
+      (* Add the constructors in the signature. *)
+      let add_cons :
+            sig_state * sym list -> ident * p_term -> sig_state * sym list
+        = fun(ss, cons_list) (id, a) ->
+        let (ss, sym_cons) = handle_symbol ss e Const id [] a in
+        (ss, sym_cons::cons_list)
+      in
+      let (ss, cons_list) = List.fold_left add_cons (ss, []) l in
+      (* Compute the induction principle *)
+      let ind_typ = Inductive.principle ss cmd.pos sym_typ cons_list in
+      (* Add the induction principle in the signature *)
+      let ind_name = Pos.make cmd.pos ("ind_" ^ id.elt) in
+      Sign.add_ident (ss.signature) ("ind_" ^ id.elt);
+      let (ss, sym_ind) =
+        Sig_state.add_symbol ss e Defin ind_name ind_typ [] None
+      in
+      Sign.add_inductive ss.signature sym_typ cons_list sym_ind None;
+      (* Add the rules of induction principle in the signature *)
+      let rs = Inductive.ind_rule ss cmd.pos sym_typ in
+      let (ss, ind_rules) = handle_rules ss rs       in
+      (* Store inductive structure in the field "sign_ind" of the signature *)
+      Sign.add_inductive ss.signature sym_typ cons_list sym_ind
+        (Some ind_rules);
```

Comment:
Why Sign.add_inductive needs to take rules as argument? Rules are recorded in the type sym itself.

Reply:
Yes your solution is better! I didn't see this.

Reply:
Should I use `add_rule` ?
Can't I add the new symbol with its rules directly?
I've tried to do this but it doesn't work...

Reply:
add_symbol creates a symbol with no rule. Sign.add_rule is necessary but sufficient as it does not update the decision tree. You need to look at what is done in handle_cmd, case P_rules. For each p_rule, we scope getting a pre_rule, we check SR getting a rule, we call Sign.add_rule. Once we finished to add all the rules, we call Tree.update_dtree on the list of symbols defined by the added rules. Something similar is done for unification rules in the case P_set/P_config_unif_rule.
Note that, in our case, Sign.add just do `sym.sym_rules := !(sym.sym_rules) @ [r]` which is not very efficient but is done just for the option `--keep-rule-order` (@gabrielhdt we could improve this by checking Tree.rule_order first). Anyway, in our case, it does not matter and we could bypass Sign.add and directly do  `sym.sym_rules := r :: !(sym.sym_rules)`.

Reply:
@gabrielhdt Ah but it may be the case that a file A.lp is compiled without the option --keep-order and then a file B.lp depending on A.lp is compiled with the option --keep-order. Then, what is the result? The doc just says: "forces the rewriting engine to try and apply the rules in the order of their declaration". There seems to be something wrong with this option. Gabriel, what do you think?

Reply:
That is indeed a good remark, the `--keep-order` changes the decision trees (introducing nodes that check if the stack is exhausted). It may be the case that importing without `--keep-order` a file compiled with `--keep-order` is safe, but importing with `--keep-order` a file that has been compiled without the flag is likely to lead to runtime failures.

We suggested in an issue #331 to declare the keep order as a flag related to symbols, which would seem safe.